client/packages/common/src/api/QueryClientProvider.tsx

```diff
@@ -0,0 +1,34 @@
+import React, { FC, PropsWithChildren } from 'react';
+import { PropsWithChildrenOnly } from '@common/types';
+import {
+  QueryClient,
+  QueryClientProvider as ReactQueryClientProvider,
+} from 'react-query';
+import { createRegisteredContext } from 'react-singleton-context';
+
+const QueryClientContext = createRegisteredContext<{ client: QueryClient }>(
+  'query-client',
+  { client: {} as QueryClient }
+);
+
+export const QueryClientProvider: FC<
+  PropsWithChildren<{ client: QueryClient }>
+> = ({ client, children }) => {
+  return (
+    <QueryClientContext.Provider value={{ client }}>
+      <QueryClientProviderProxy>{children}</QueryClientProviderProxy>
+    </QueryClientContext.Provider>
+  );
+};
+
+export const QueryClientProviderProxy: FC<PropsWithChildrenOnly> = ({
```

Comment:
Was wondering, is this just a stylistic preference or am i missing something, vs:

```react
export const QueryClientProvider: FC<
  PropsWithChildren<{ client: QueryClient }>
> = ({ client, children }) => {
  return (
    <QueryClientContext.Provider value={{ client }}>
      <ReactQueryClientProvider client={client}>
         {children}
      </ReactQueryClientProvider>
    </QueryClientContext.Provider>
  );
};

```

Reply:
no, not a stylistic preference. the plugin needs to implement the Context, while also having access to the `client` which is initialised in the host app.

To do that, I've had to use the `react-singleton-context` which stores the `client`. However, existing components (from the packages, such as QueryClient) are looking for their own Context/Provider - and will not 'see' the custom one. So, I've had to store the state in the higher level, then fetch the state from that higher level in the Proxy which then instantiates the type of Provider that the package is trying to find.